Update docs for 2.0 (#352)
sunpy/ndcube: docs/visualization.rst

.. code-block:: python

  >>> my_cube.plot(plot_axes=[..., 'y', 'x']) # doctest: +SKIP

`~ndcube.NDCube.plot` uses `~astropy.visualization.wcsaxes.WCSAxes` to produce all plots.  This enables a rigorous representation of the coordinates on the plot, including those that are not aligned to the pixel grid.  It also enables the coordinates along the plot axes to be updated between frames of an animation. `ndcube.NDCube.plot` therefore allows users to decide which WCS object to use, either `~ndcube.NDCube.wcs` or `~ndcube.NDCube.combined_wcs` which also includes the `~ndcube.ExtraCoords`.  In principle, another third-part WCS can be used so long as it is a valid description of all array axes.

.. code-block:: python

>   >>> my_cube.plot(wcs=my_cube.combined_wcs)   # doctest: +SKIP

Reviewer: it would be nice in general if these actually plot on the docs? just to get a better idea

Reviewer 2: See https://matplotlib.org/sampledoc/extensions.html#inserting-matplotlib-plots for how to do that.

  >>> mask[-1, -1][:] = True  # Now mask some values.
  >>> # Define uncertainty, metadata and unit.
  >>> uncertainty = StdDevUncertainty(np.sqrt(np.abs(data)))
  >>> meta = {"Description": "This is example NDCube metadata."}
  >>> unit = u.ct

  >>> # Instantiate NDCube with supporting data.
  >>> my_cube = NDCube(data, wcs=wcs, uncertainty=uncertainty, mask=mask, meta=meta, unit=unit)

The plot method can be called very simply.

.. plot::
  :context:
  :include-source:

  >>> import matplotlib.pyplot as plt
  >>> ax = my_cube.plot()
  >>> plt.show()

For a >2 array axes, as is the case above, an animation object is returned displaying either a line or image with sliders for each additional array axis.
These sliders are used to sequentially update the line or image as it moves along its corresponding array axis, thus animating the data.
By default and image animation is returned.
(See below to learn how to use ``plot_axes`` kwarg to produce a line animation.)

For for data with two array axes, an image is produced similar to that of `matplotlib.pyplot.imshow`.

.. plot::
  :context:
  :include-source:

  >>> ax = my_cube[0].plot()
  >>> plt.show()

> Note that the bottom row of the image is empty because those array elements are marked ``True`` is ``.mask``.

Reviewer 2: it isn't empty??

Author: No, I've defined the mask in the expanding code block above to demonstrate its affect.

> Note that the bottom row is of the image is empty because those array elements are marked ``True`` is ``.mask``.

Reviewer: ```suggestion
Note that the bottom row of the image is empty because those array elements are marked ``True`` is ``.mask``.
```

For data with one array axis, a line plot is produced, similar to `matplotlib.pyplot.plot`.

.. plot::
  :context:
  :include-source:

  >>> ax = my_cube[1, 1].plot()
  >>> plt.show()

Setting the x and y ranges of the plot can be done simply by indexing the `~ndcube.NDCube` object to the desired region of interest and then calling the plot method, e.g.

.. plot::
  :context:
  :include-source:

  >>> ax = my_cube[0, 1:3, 1:4].plot()
  >>> plt.show()

.. plot::
  :nofigs:

  >>> plt.clf()
  >>> plt.cla()
  >>> plt.close()

> No args are required.
> The necessary information to generate the plot is derived from the data and metadata in the `~ndcube.NDCube`.

Reviewer 2: ```suggestion
Note how no arguments are required, the necessary information to generate the plot is derived from the data and metadata in the `~ndcube.NDCube`.
```

However optional keywords enable customization of the visualization.

> For `~ndcube.NDCube` instances with more than one array axis, the ``plot_axes`` keyword is used to determine which array axes are displayed on which plot axes.It is set to a list with a length equal to the number of array axes.

Reviewer 2: ```suggestion
For `~ndcube.NDCube` instances with more than one array axis, the ``plot_axes`` keyword is used to determine which array axes are displayed on which plot axes.
It is set to a list with a length equal to the number of array axes.
```

If no ``'y'`` axis is provided, a line animation is produced.
By default the ``plot_axes`` argument is set so that the last array axis to shown on the x-axis and the penultimate array axis is shown on the y-axis.

.. plot::
  :context:
  :include-source:

  >>> ax = my_cube.plot(plot_axes=[None, 'x', 'y'])
  >>> plt.show()

.. plot::
  :nofigs:

  >>> plt.clf()
  >>> plt.cla()
  >>> plt.close()

The ``plot_axes`` kwarg can also be used to generated a line animation by omitting the ``'y'`` entry.

.. plot::
  :context:
  :include-source:

  >>> ax = my_cube.plot(plot_axes=[None, None, 'x'])
  >>> plt.show()

.. plot::
  :nofigs:

  >>> plt.clf()
  >>> plt.cla()
  >>> plt.close()

`~ndcube.NDCube.plot` uses `~astropy.visualization.wcsaxes.WCSAxes` to produce all plots.
This enables a rigorous representation of the coordinates on the plot, including those that are not aligned to the pixel grid.
It also enables the coordinates along the plot axes to be updated between frames of an animation.
`ndcube.NDCube.plot` therefore allows users to decide which WCS object to use, either `~ndcube.NDCube.wcs` or `~ndcube.NDCube.combined_wcs` which also includes the `~ndcube.ExtraCoords`.

> In principle, another third-part WCS can be used so long as it is a valid description of all array axes.

Reviewer 2: ```suggestion
```

.. plot::
  :context:
  :include-source:

  >>> ax = my_cube.plot(wcs=my_cube.combined_wcs)
  >>> plt.show()

.. plot::
  :nofigs:

  >>> plt.clf()
  >>> plt.cla()
  >>> plt.close()

Visualizing NDCubeSequences
===========================
Since ndcube 2.0, the `~ndcube.NDCubeSequence` visualization support has been dropped.
The rationale for this is outlined in `Issue #321 <https://github.com/sunpy/ndcube/issues/321>`_ on the ndcube GitHub repo.
If you feel that `~ndcube.NDCubeSequence` visualization should be supported again, please let us know by commenting on that issue and telling us of your use case.

> Better still, if you would like to work on the infrastructure required to support `~ndcube.NDCubeSequence` visualization is a post ndcube 2.0 world let us know by commenting on the issue.

Reviewer: ```suggestion
Better still, if you would like to work on the infrastructure required to support `~ndcube.NDCubeSequence` visualization in a post ndcube 2.0 world let us know by commenting on the issue.
```

Despite this the lack of `~ndcube.NDCubeSequence` visualization support, you can still visualize the data in `~ndcube.NDCubeSequence` in a number of ways.
You can slice out a single `~ndcube.NDCube` and use its `~ndcube.NDCube.plot` method.
You can extract the data and use the myriad of plotting packages available in the Python ecosystem.
Finally, if you want to be advanced, you can write your own mixin class to define the plotting methods.
Below, we will outline these latter two options in a little more detail.

Extracting and Plotting NDCubeSequence Data with Matplotlib
-----------------------------------------------------------
In order to produce plots (or perform other analysis) outside of the ``ndcube`` framework, it may be useful to extract the data from the `~ndcube.NDCubeSequence` into single `~numpy.ndarray` instances.

> Let's first define an `~ndcube.NDCubeSequence` with a common axis of 0 and and time as an extra coord stretching across the cube along the common axis.

Reviewer: ```suggestion
Let's first define an `~ndcube.NDCubeSequence` with a common axis of 0 and time as an extra coord stretching across the cube along the common axis.
```

    >>> wavelength = spectrum_sequence[0].axis_world_coords("em.wl")[0]
    >>> # As the times may not be uniform, we can use NonUniformImage to show non-uniform pixel sizes.
    >>> fig, ax = plt.subplots(1, 1)  # doctest: +SKIP
    >>> im = mpl.image.NonUniformImage(
    ...     ax, extent=(times[0], times[-1], wavelength[0], wavelength[-1]))  # doctest: +SKIP
    >>> im.set_data(wavelength, times.mjd, intensity)  # doctest: +SKIP
    >>> ax.add_image(im)  # doctest: +SKIP
    >>> ax.set_xlim(times.mjd[0], times.mjd[-1])  # doctest: +SKIP
    >>> ax.set_xlabel("Time [Modified Julian Day]")  # doctest: +SKIP
    >>> ax.set_ylim(wavelength[0].value, wavelength[-1].value)  # doctest: +SKIP
    >>> ax.set_ylabel(f"Wavelength [{wavelength.unit}]")  # doctest: +SKIP
    >>> plt.show()  # doctest: +SKIP

Now let's say we want to animate our data, for example, show how the intensity changes over wavelength and time.
For this we can use `~ndcube.visualization.animator.ImageAnimator`.
This class is not well suited to displaying the complex relationship between coordinates that we are used to with `~astropy.visualization.wcsaxes.WCSAxes`.

> For example, non-linear coordinates non-independent coordinates.

Reviewer: ```suggestion
For example, non-linear coordinates and non-independent coordinates.
```
